vispy/visuals/_scalable_textures.py

```diff
@@ -0,0 +1,441 @@
+            data /= clim[1] - clim[0]
+        if should_cast_to_f32(data.dtype):
+            data = data.astype(np.float32)
+        return data
+
+    def scale_and_set_data(self, data, offset=None, copy=True):
+        """Upload new data to the GPU, scaling if necessary."""
+        self._data_dtype = data.dtype
+
+        clim = self._clim
+        is_auto = isinstance(clim, str) and clim == 'auto'
+        if data.ndim == self._ndim or data.shape[self._ndim] == 1:
+            if is_auto:
+                clim = np.min(data), np.max(data)
+            clim = (np.float32(clim[0]), np.float32(clim[1]))
+            data = self._scale_data_on_cpu(data, clim, copy=copy)
+            data_limits = clim
+        else:
+            data_limits = self._get_default_clims(data)
+            if is_auto:
+                clim = data_limits
+
+        self._clim = clim
+        self._data_limits = data_limits
+        return super().scale_and_set_data(data, offset=offset, copy=copy)
+
+
+class GPUScaledTextureMixin(_ScaledTextureMixin):
+    """Texture class for smarter scaling and internalformat decisions.
+
+    This texture class uses internal formats that are not supported by
+    strict OpenGL 2/ES drivers without additional extensions. By using
+    this texture we upload data to the GPU in a format as close to
+    the original data type as possible (32-bit floats on the CPU are 32-bit
+    floats on the GPU). No normalization/scaling happens on the CPU and
+    all of it happens on the GPU. This should avoid unnecessary data copies
+    as well as provide the highest precision for the final visualization.
+
+    The texture format may either be a GL enum string (ex. 'r32f'), a numpy
+    dtype object (ex. np.float32), or 'auto' which means the texture will
+    try to pick the best format for the provided data. By using 'auto' you
+    also give the texture permission to change formats in the future if
+    new data is provided with a different data type.
+
+
+    This class should only be used internally. For similar features where
+    scaling occurs on the CPU see
+    :class:`vispy.visuals._scalable_textures.CPUScaledTextureMixin`.
+
+    To use this mixin, a subclass should be created to combine this mixin with
+    the texture class being used. Existing subclasses already exist in this
+    module. Note that this class **must** appear first in the subclass's parent
+    classes so that its ``__init__`` method is called instead of the parent
+    Texture class.
+
+    """
+
+    # dtype -> internalformat
+    # 'r' will be replaced (if needed) with rgb or rgba depending on number of bands
+    _texture_dtype_format = {
+        np.float32: 'r32f',
+        np.float64: 'r32f',
+        np.uint8: 'r8',
+        np.uint16: 'r16',
+        # np.uint32: 'r32ui',  # not supported texture format in vispy
+        np.int8: 'r8',
+        np.int16: 'r16',
+        # np.int32: 'r32i',  # not supported texture format in vispy
```

Comment:
Why are these not supported? Is this a matter of adding a field somewhere in gloo?

Reply:
Ah: https://github.com/vispy/vispy/blob/a20297a34bf5de3204111842b0d77f09f7a59500/vispy/gloo/texture.py#L68-L76